### PR TITLE
Round stored flags

### PR DESCRIPTION
**Pull request by roystgnr**
This fixes the robustness issues discussed in https://github.com/idaholab/moose/issues/6621, and updates the cycles_per_step regression test accordingly.

### CODE REVIEW

File: framework/src/base/FlagElementsThread.C [hunk 1]
```diff
@@ -57,7 +57,10 @@ void
 FlagElementsThread::onElement(const Elem *elem)
 {
   dof_id_type dof_number = elem->dof_number(_system_number, _field_var_number, 0);
-  Marker::MarkerValue marker_value = (Marker::MarkerValue)_serialized_solution[dof_number];
+
+  // This is basically lround() but doesn't require C99 or C++11
+  Marker::MarkerValue marker_value =
+    (Marker::MarkerValue)(std::ceil(_serialized_solution[dof_number]-0.5));
```
Comment (friedmud) [inline framework/src/base/FlagElementsThread.C]: Still against this.  But others can outvote me.

BTW: We can at least require C99... right?

**Comment by permcody** [comment]
I don't like the look of this but I also understand and appreciate the dilemma. John is currently on vacation (which is why I'm not tagging him). I think we'll wait for him to get back to weigh in on this a bit too.

**Comment by roystgnr** [comment]
Should I just skip the intermediate step here, and go full-hog to the "take away the rounding but assert that rounding would have done the same as truncation" phase?  That won't pass tests unless I also add a libMesh update to include https://github.com/libMesh/libmesh/pull/871 , but you guys are nearly up to date with libMesh master right now so the update won't be a huge change.

**Comment by roystgnr** [comment]
My preference would be to merge this now (for loose values of "now" which include "after John gets back and casts his vote"), kick the Civet tires on and then merge libMesh/libmesh#871 (my own runs indicate that the remaining test failure there will be fixed by the updated gold solution here, but I'd feel safer knowing your CI agrees), then merge a Moose PR with the libMesh update plus rounding->truncate+assert change.

**Comment by permcody** [comment]
Are you in a hurry to get this merged? I'd like John to weigh in a bit with this when he gets back from his spring break vacation. We can just leave this for a few more days if that's ok with you.

**Comment by roystgnr** [comment]
Delay is fine by me.  @bwspenc is the most urgent user of the libMesh optimizations, but (now that I've debugged it...) he ought to be fine working off the branch.

**Comment by bwspenc** [comment]
I'm in no rush. I haven't had a chance to test the optimizations @roystgnr made, and won't until next week.

File: framework/src/base/FlagElementsThread.C (hunk 1 above)
Comment (jwpeterson) [inline framework/src/base/FlagElementsThread.C]: We already use C99 `round` in GeneratedMesh.C and there's a `MathUtils::round()` interface which also uses it.  The only difference between `round` and `lround` is that the latter returns a `long int`... so if it's the right thing here we should definitely use it.

Comment (jwpeterson) [inline framework/src/base/FlagElementsThread.C]: Oh, I guess `lround` may also raise a floating-point exception is outside the range... but this shouldn't be an issue for marker flags.

Comment (jwpeterson) [inline framework/src/base/FlagElementsThread.C]: Let's also do away with the C-style cast while we're at it?

**Comment by permcody** [comment]
This is relevant to Jesse Carter's issue too
Before:
https://mail.google.com/mail/u/0?ui=2&ik=fefb650270&attid=0.3&th=153bb2da0deb46cc&view=fimg&realattid=06b9988d-d3c4-45b1-b68c-dd18b64ae184&disp=thd&attbid=ANGjdJ_SVkqypz9Ly-BkUdDHY_wcKZgSBLUTaj8nSvou2i47I4IG4OzrwWmfXCm2z7qlhjrCHgqRz8tQSFb9UCbyzAz03548p9IXfQ1zNrrDBvMJPYuLQsfT9TBRKg0&ats=2524608000000&sz=w1256-h1341

After:
adaptivity_with_projection

**Comment by jwpeterson** [comment]
After reading through the associated issue as well, I think @friedmud is saying that:

1. "way too much of a pain to track down every place where that is done in the code."
2. "rounding is that it can cover up real mistakes."

For 1, is there anything else besides Markers which are supposed to be integer fields?  I suppose this could also be an issue for the `ProcessorIDAux` AuxKernel?  Anyway, this will (eventually) be fixed by using the INTEGER FEType I think.

For 2, at least in FlagElementsThread.C, we can just check for actual mistakes in the meantime until INTEGER is implemented.  std::modf is a nifty little function that splits an input value into integral and fractional parts, then we just check that the fractional part is below some tolerance before accepting the integer value.  

@roystgnr I'd like to take care of this today if possible, so that libMesh/libmesh#871 can be merged.  Unless you have any objections I'll pull your branch and change it slightly to check for errors while doing rounding of the Markers?

**Comment by roystgnr** [comment]
@jwpeterson, go right ahead.  My only quibble: rather than use modf, I'd suggest simply testing that rounded_value == original_value.  That should avoid overhead in opt mode and should be more readable.

**Comment by jwpeterson** [comment]
> That should avoid overhead in opt mode and should be more readable.

Yes, the modf thing idea was dumb, e.g. if you had 2.999, your fractional part is nearly 1...

**Pull request merged**